### PR TITLE
Datatools

### PR DESCRIPTION
**Pull request by tomlepaine**
A few tools to simple data generation:
- `sequential` : Generate sequential indices.
- `random` : Shuffle indices then draw samples without replacement until all samples are drawn, then shuffle and repeat.
- `batch_generator` : Draw samples from the data to create batches.
- `async_batch_generator` : Asynchronously draw samples from the data to create batches. This can be useful if `data.get(samples)` takes a long time (for instance if it is reading samples from disk), or if `batch_function` takes a long time (for instance image augmentation).
- `NumpyData` : A simple data backend. Works with `numpy` ndarrays, but should also work with `HDF5` files on disk.

The additional tools are mostly private utils used by the above tools:
- `queue_generator`
- `BatchFetcher`

See tests for basic usage.

I will add complete doc strings. But if I could get some intermediate feedback that would be cool.

### CODE REVIEW

**Comment by tomlepaine** [comment]
It seems somehow I have a whole bunch of old commits tagging along. Anyone know how to fix this? @jrock08 ?

**Comment by jrock08** [comment]
git rebase origin/master instead of git merge origin/master

File: aiutils/data/data.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def sequential(batch_size, num_samples, num_epochs=1):
+    """Generate sequence indices.
+    """
+    for epoch in range(num_epochs):
+        indices = np.arange(num_samples)
+        for i in range(0, num_samples - batch_size + 1, batch_size):
+            slice_ = slice(i, i + batch_size)
+            yield indices[slice_]
```
Comment (BigRedT) [inline aiutils/data/data.py]: Impressive use of Generators in python!

File: aiutils/data/data.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+"""Utility functions for making batch generation easier."""
+import numpy as np
+from multiprocessing import Process, Queue
+
+
+def sequential(batch_size, num_samples, num_epochs=1):
```
Comment (BigRedT) [inline aiutils/data/data.py]: sequential is a little too general....change name to something that shows that its is generating indices

File: aiutils/data/data.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        indices = np.random.permutation(num_samples)
+        for i in range(0, num_samples - batch_size + 1, batch_size):
+            slice_ = slice(i, i + batch_size)
+            yield indices[slice_]
+
+
+def batch_generator(data, index_generator, batch_function=None):
+    """Generate batches of data.
+    """
+    for samples in index_generator:
```
Comment (BigRedT) [inline aiutils/data/data.py]: sample_ids instead of samples

File: aiutils/data/data.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        batch = data.get(samples)
```
Comment (BigRedT) [inline aiutils/data/data.py]: Probably you were going to do it already but make sure user knows that this function expects data to have get() function

File: aiutils/data/data.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        value = queue.get()
+        if value is not sentinel:
+            yield value
+        else:
+            return
+
+
+class BatchFetcher(Process):
+    def __init__(self, queue, batch_generator):
+        super(BatchFetcher, self).__init__()
```
Comment (BigRedT) [inline aiutils/data/data.py]: i know python doc recommends super instead of Process.__init__() but i am not very sure about why?

Comment (jrock08) [inline aiutils/data/data.py]: http://stackoverflow.com/questions/576169/understanding-python-super-with-init-methods

File: aiutils/data/data.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    batcher = batch_generator(data, index_generator, batch_function)
+
+    queue = Queue(maxsize=queue_maxsize)
+    fetcher = BatchFetcher(queue, batcher)
+    fetcher.start()
+
+    queue_batcher = queue_generator(queue)
+    return queue_batcher
+
+
```
Comment (BigRedT) [inline aiutils/data/data.py]: I haven't used multiprocessing before but would the following work -
```
def write_to_queue(queue, batch_generator):
    for batch in batch_generator:
        queue.put(batch)
    queue.put(None)
```
And then use as 
```
batch_enque_process = Process(target=write_to_queue, args=(queue, batcher))
```

File: aiutils/data/data.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+class NumpyData(object):
+    def __init__(self, array):
+        self.array = array
+
+    def get(self, indices):
+        if not isinstance(indices, np.ndarray):
+            indices = np.array(indices)
+        return self.array[indices]
```
Comment (BigRedT) [inline aiutils/data/data.py]: are you assuming that array is a 1d array?

**Comment by BigRedT** [comment]
@tomlepaine I have added comments to your code. I will look into the test units once you have replied to these comments. Nice code btw, I learned a lot about using generators and multiprocessing module from your code.

File: aiutils/data/data.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    for epoch in range(num_epochs):
+        indices = np.arange(num_samples)
+        for i in range(0, num_samples - batch_size + 1, batch_size):
+            slice_ = slice(i, i + batch_size)
+            yield indices[slice_]
+
+
+def random(batch_size, num_samples, num_epochs=1):
+    """Generate random indices.
+    """
```
Comment (kevjshih) [inline aiutils/data/data.py]: I think it would be more straightforward to directly call np.random.shuffle

File: aiutils/data/data.py (hunk 5 above)
Comment (tomlepaine) [inline aiutils/data/data.py]: Tl;dr. Is super better or worse 😜

File: aiutils/data/data.py (hunk 6 above)
Comment (tomlepaine) [inline aiutils/data/data.py]: Hmm. Fair enough. I learned it from Ross Girshick's `py-faster-rcnn` code ([see here]).(https://github.com/rbgirshick/py-faster-rcnn/blob/96dc9f1dea3087474d6da5a98879072901ee9bf9/lib/roi_data_layer/layer.py#L161)

I thought it was easy to understand from a high level. When you inherit from `Process`, `run` runs in its own process 😈 .

File: aiutils/data/data.py (hunk 7 above)
Comment (tomlepaine) [inline aiutils/data/data.py]: I found out this works for when `array` is any rank. It always indexes on the first dimension. So for 2d arrays it would be equivalent to: `array[indices, :]`

File: aiutils/data/data.py (hunk 8 above)
Comment (tomlepaine) [inline aiutils/data/data.py]: You mean instead of random?

So I agree that would work, but then you have to decide when to shuffle, and how many times to shuffle. I don't see why you should re-write that each time, when all you want to do is grab random batches.

**Comment by tomlepaine** [comment]
Thanks @BigRedT. I've been busy with research and a side project. I'll address these comments and add docstrings soon.

**Comment by tomlepaine** [comment]
@jrock08 Ooo, that is way cleaner. Thanks man.

File: aiutils/data/data.py (hunk 2 above)
Comment (tomlepaine) [inline aiutils/data/data.py]: Good point. It should be more specific. How about `sequential_generator` and `random_generator`?

File: aiutils/data/data.py (hunk 4 above)
Comment (tomlepaine) [inline aiutils/data/data.py]: Yeah, that is definitely an important thing to mention in the docstrings.

**Comment by kevjshih** [comment]
I think this code still truncates the dataset if the dataset size isn't divisible by the batch size. Unless TF still has issues with variable batch sizes, we should probably return the remaining training examples in a smaller batch instead.

File: aiutils/data/data.py (hunk 2 above)
Comment (BigRedT) [inline aiutils/data/data.py]: that or `sequential_indices` whatever you prefer

File: aiutils/data/data.py (hunk 6 above)
Comment (BigRedT) [inline aiutils/data/data.py]: yeah but you need to understand that `run()` gets called whenever you call `start()` which may not be that big a deal. But even if you want to do it this way I would recommend naming the class something that indicates you are writing to queue. I guess `BatchFetcher` was a slightly confusing name.

File: aiutils/data/data.py (hunk 7 above)
Comment (BigRedT) [inline aiutils/data/data.py]: Oh good to know

File: aiutils/data/data.py (hunk 5 above)
Comment (jrock08) [inline aiutils/data/data.py]: Super deals with multi-inheritance

File: aiutils/data/data.py (hunk 8 above)
Comment (jrock08) [inline aiutils/data/data.py]: Change range to xrange in case num_epochs is large

File: aiutils/data/data.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        indices = np.arange(num_samples)
+        for i in range(0, num_samples - batch_size + 1, batch_size):
+            slice_ = slice(i, i+batch_size)
+            yield indices[slice_]
+
+
+def random(batch_size, num_samples, num_epochs=1):
+    """Generate random indices.
+    """
+    for epoch in range(num_epochs):
```
Comment (jrock08) [inline aiutils/data/data.py]: Change range to xrange  (recommend doing this everywhere actually)

File: aiutils/data/data.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    """Generate sequence indices.
+    """
+    for epoch in range(num_epochs):
+        indices = np.arange(num_samples)
+        for i in range(0, num_samples - batch_size + 1, batch_size):
+            slice_ = slice(i, i+batch_size)
+            yield indices[slice_]
+
+
+def random(batch_size, num_samples, num_epochs=1):
```
Comment (jrock08) [inline aiutils/data/data.py]: Do you think you could allow for num_epochs=-1 which would create an unbounded number of epoch iterations.

a way to do this is
`for i in xrange(n) if n!=-1 else itertools.count():`

File: aiutils/data/data.py (hunk 9 above)
Comment (kevjshih) [inline aiutils/data/data.py]: might as well use np.arange since it's already imported
http://stackoverflow.com/questions/10698858/built-in-range-or-numpy-arange-which-is-more-efficient

xrange won't work in python3 anyways.

Comment (kevjshih) [inline aiutils/data/data.py]: Actually that's a bad idea if you're using a for loop. Just read the whole thing I linked. I think there's a from builtins import range solution.

Comment (jrock08) [inline aiutils/data/data.py]: To clarify I agree with Kevin use:

`from builtins import range`

File: aiutils/data/data.py (hunk 10 above)
Comment (kevjshih) [inline aiutils/data/data.py]: This could be useful for small datasets where each epoch finishes in a matter of minutes but you'd like to train for a week without specifying an arbitrarily large number.

File: aiutils/data/data.py (hunk 8 above)
Comment (kevjshih) [inline aiutils/data/data.py]: Sorry, missed the part about this being indices. LGTM.

**Comment by BigRedT** [comment]
@kevjshih: probably truncation should be handled in the `get()` function in `data` class rather than in this
